pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_add_constraint.go

```diff
@@ -38,3 +75,240 @@ func alterTableAddConstraint(
+		panic(err)
+	} else if skip {
+		return
+	}
+
+	// 2. CheckDeepCopy whether this check constraint is syntactically valid. See the comments of
+	// DequalifyAndValidateExprImpl for criteria.
+	ckDef := t.ConstraintDef.(*tree.CheckConstraintTableDef)
+	ckExpr, _, colIDs, err := schemaexpr.DequalifyAndValidateExprImpl(b, ckDef.Expr, types.Bool, "CHECK",
+		b.SemaCtx(), volatility.Volatile, tn,
+		func() colinfo.ResultColumns {
+			return getNonDropResultColumns(b, tbl.TableID)
+		},
+		func(columnName tree.Name) (exists bool, accessible bool, id catid.ColumnID, typ *types.T) {
+			return columnLookupFn(b, tbl.TableID, columnName)
+		},
+	)
+	if err != nil {
+		panic(err)
+	}
+	typedCkExpr, err := parser.ParseExpr(ckExpr)
+	if err != nil {
+		panic(err)
+	}
+
+	// 3. Add relevant check constraint element: CheckConstraint and ConstraintName.
+	constraintID := b.NextTableConstraintID(tbl.TableID)
+	b.Add(&scpb.CheckConstraint{
+		TableID:               tbl.TableID,
+		ConstraintID:          constraintID,
+		ColumnIDs:             colIDs.Ordered(),
+		Expression:            *b.WrapExpression(tbl.TableID, typedCkExpr),
+		FromHashShardedColumn: ckDef.FromHashShardedColumn,
+	})
+
+	constraintName := string(ckDef.Name)
+	if constraintName == "" {
+		constraintName = generateUniqueCheckConstraintName(b, tbl.TableID, ckDef.Expr)
+	}
+	b.Add(&scpb.ConstraintName{
+		TableID:      tbl.TableID,
+		ConstraintID: constraintID,
+		Name:         constraintName,
+	})
+}
+
+// validateConstraintNameIsNotUsed checks that the name of the constraint we're
+// trying to add isn't already used, and, if it is, whether the constraint
+// addition should be skipped:
+// - if the name is free to use, it returns false;
+// - if it's already used but IF NOT EXISTS was specified, it returns true;
+// - otherwise, it returns an error.
+// TODO (xiang): It is written knowing that we will only see check constraint
+// for now. Complete it by also considering all other types of constraints as well.
+func validateConstraintNameIsNotUsed(
+	b BuildCtx, tn *tree.TableName, tbl *scpb.Table, t *tree.AlterTableAddConstraint,
+) (skip bool, err error) {
+	var name tree.Name
+	var ifNotExists bool
+	switch d := t.ConstraintDef.(type) {
+	case *tree.CheckConstraintTableDef:
+		name = d.Name
+		ifNotExists = d.IfNotExists
+	case *tree.ForeignKeyConstraintTableDef:
+		panic(scerrors.NotImplementedErrorf(t, "FK constraint %v not yet implemented", d.Name))
+	case *tree.UniqueConstraintTableDef:
+		panic(scerrors.NotImplementedErrorf(t, "UNIQUE constraint %v not yet implemented", d.Name))
+	default:
+		return false, errors.AssertionFailedf(
+			"unsupported constraint: %T", t.ConstraintDef)
+	}
+
+	if name == "" {
+		// A name was not specified; We can give it any name later.
+		return false, nil
+	}
+
+	var isInUse bool
+	scpb.ForEachConstraintName(b.QueryByID(tbl.TableID), func(
+		_ scpb.Status, _ scpb.TargetStatus, e *scpb.ConstraintName,
+	) {
+		if e.Name == name.String() {
```

Comment:
I think you're supposed to use `string(name)` here or else names which require quotes won't work.

Reply:
fixed